network/sender.go

```diff
@@ -65,6 +65,7 @@ Loop:
 		if err != nil {
 			log.Error().Err(err).Msg("could not write message")
 			rep.Failure(address)
+			eventMgr.Disconnected(address)
```

Comment:
You really need to think about where to call the disconnected event in order to have it happen only once, exactly once, but in every case. The way the shutdown cascade works is a bit specific.